### PR TITLE
feat(inventory): add per-request dsn_from for static targets

### PR DESCRIPTION
**Pull request by Kiran01bm**
## Summary

Static `target_resolver.targets` can now assemble a namespace-free MySQL DSN at
runtime from separate config and password secret references (`dsn_from`),
resolved fresh on every request. This lets a data-plane worker read a rotated
target credential without a restart, instead of relying on a pre-resolved DSN
string baked in at startup.

## What

- Add `dsn_from` to a static target (`config_ref`, `config_paths` for
  host/port, `username`, `password_ref`, optional `params`), mutually exclusive
  with `dsn`.
- `dsn_from` targets are resolved per `ResolveTarget` call, reusing the existing
  MySQL connection assembler and the fail-closed secret-ref credential resolver.
- The assembled DSN is namespace-free: the config document's database key, if
  any, is ignored (the per-operation request supplies the schema), and a DSN
  carrying a database name is still rejected.
- Existing `dsn:` targets are unchanged — still resolved once at construction.
- MySQL-only for now; a `dsn_from` target with any other type is rejected at
  construction.

## Why

A static `dsn:` reference is resolved once when the resolver is built, so a
rotated credential is only picked up after a restart. `dsn_from` mirrors the
per-use resolution already used for storage/environment DSNs and the dynamic
resolver path, making runtime credential rotation the safe default.

## Before / after

BEFORE — static target supports only dsn:
```
NewStaticResolver (once, at startup)
dsn: → secrets.Resolve → cached DSN ──┐
▼
ResolveTarget(req) ───────────► return cached DSN
rotated credential ⇒ stale until pod restart
```
AFTER — static target also supports dsn_from:
```
NewStaticResolver (once, at startup)
dsn:      → secrets.Resolve → cached DSN     (unchanged)
dsn_from: → validate shape only (no secret read yet)

ResolveTarget(req)
dsn:      → return cached DSN                (unchanged)
dsn_from: → resolve fresh, every request:
config_ref   → host/port  (dbname ignored)
password_ref → password   (fail-closed on empty)
→ MySQL assembler → namespace-free DSN
rotated credential ⇒ picked up on next request, no restart
```

**NOTE:**
Making the client cache DSN-aware — and memoizing the per-call Secrets Manager
client while threading the caller's context so a per-request secret fetch honors
the request deadline instead of a fixed 10s background timeout — is a tracked
follow-up, out of scope for this resolver-layer change.

### CODE REVIEW

File: pkg/inventory/static.go [hunk 1]
```diff
@@ -96,12 +173,24 @@ func resolveStaticTarget(target string, entry StaticTarget) (*Target, error) {
...
+	return prepared, nil
+}
+
+// resolveDSN returns the entry's DSN: the cached value for a dsn entry, or a
+// freshly assembled namespace-free DSN for a dsn_from entry.
+func (e staticTargetEntry) resolveDSN(ctx context.Context, req Request) (string, error) {
+	if e.dsn != nil {
+		return *e.dsn, nil
+	}
+	dsn, err := e.dsnFrom.assemble(ctx, req, e.databaseType)
```
[the remark this answers is not in the document]
Comment (Kiran01bm) [reply pkg/inventory/static.go:199]: thanks - the PR is resolver-layer only: `dsn_from` now re-resolves on every `ResolveTarget` - restart-free rotation isn't delivered here because `TargetRouter` caches `LocalClient by {target,type,environment,database} `(DSN omitted) and `LocalClient` pins the DSN at construction.

File: pkg/inventory/static.go [hunk 2]
```diff
@@ -121,3 +210,115 @@ func validateResolvedStaticTargetDSN(target, databaseType, dsn string) error {
...
+		if v != float64(int64(v)) {
+			return "", fmt.Errorf("key %q must be an integer", key)
+		}
+		return strconv.FormatInt(int64(v), 10), nil
+	case json.Number:
+		return v.String(), nil
+	default:
+		return "", fmt.Errorf("key %q must be a string or number", key)
+	}
+}
```
[the remark this answers is not in the document]
Comment (Kiran01bm) [reply pkg/inventory/static.go:345]: fixed now - `optionalNumericField` now trims whitespace and requires a base-10 integer in 1–65535; blank/missing still means default - also added tests for non-numeric, out-of-range, and whitespace-trimmed cases.

File: pkg/api/config_test.go [hunk 3]
```diff
@@ -164,6 +164,33 @@ databases:
...
+storage:
+  dsn: env:MYSQL_DSN
+target_resolver:
+  targets:
+    apse2-prod:
+      type: mysql
+      dsn_from:
+        config_ref: secretsmanager:/pe/aurora/schemabot-pilot-prod-apse2/spirit_credentials
+        password_ref: secretsmanager:/pe/aurora/schemabot-pilot-prod-apse2/spirit_credentials#password
+        username: m_spirit
```
[the remark this answers is not in the document]
Comment (Kiran01bm) [reply pkg/api/config_test.go]: fixed — replaced the real-ish secrets manager paths, target name, and username with fictitious placeholders (example-target, /example/schemabot/target-credentials, schemabot).

**Review by aparajon — APPROVED** [review]

**Comment by aparajon** [comment]
🤖 Correctness review (built the head, ran the inventory and api config tests with `-race` — all pass; audited against the AGENTS.md self-review bar).

**What checks out.** The feature reuses the right shared pieces rather than reinventing them: `SecretRefCredentialResolver` (fail-closed on an empty resolved password), `MySQLConnectionAssembler` (namespace-free DSN via `mysql.NewConfig`/`FormatDSN` — no string-built DSNs), and a post-assembly `ParseDSN` check that rejects any DSN carrying a database name. Mutual exclusion of `dsn`/`dsn_from` and shape validation happen at construction, while the password is deliberately not read until request time. Port validation is robust (string and float64 accepted, fractional rejected, 1–65535 range, whitespace trimmed) and every branch is tested — including the config document carrying a `database` key that must be ignored. Concurrency is safe: the target map is immutable after construction and per-request assembly has no shared state. The deletions are pure restructuring — no behavior or test removals. All three Copilot threads are addressed on the head, and the rotation caveat in the PR body is honest about the client-cache limitation.

**Findings:**

1. **The documented rotation caveat has a sharper, undocumented half: a new per-request availability dependency with zero payoff until the router follow-up.** `ResolveTarget` runs *before* the client-cache lookup in the target router (`clientForTarget`), and on a cache hit the freshly assembled DSN is **discarded** — the pinned-DSN cached client serves the request anyway. So today a `dsn_from` target makes every routed request depend on secret-backend availability and buys nothing for it: a transient Secrets Manager failure fails requests that a `dsn:` target would serve fine from cache. With `secretsmanager:` refs that's also two `GetSecretValue` calls plus two fresh AWS config/client constructions per routing call (and the natural config shape — one secret holding host/port and password — fetches the *same* secret twice). With `file:` refs (the ESO case the field docs name) both cost and risk are negligible. Ask: either document `file:` as the intended backend for `dsn_from` until the DSN-aware client cache lands, or sequence that follow-up to land close behind so the new failure surface pays for itself.

2. **The per-request secret path ignores the caller's context.** `assemble` threads `ctx` to `ResolveCredentials`, which discards it (`_ context.Context`), and the secrets-manager resolution builds its own `context.Background()` with a fixed 10s timeout — so the new per-request network I/O can't be cancelled by the request deadline; a slow secret fetch holds routing for up to 10s regardless of the caller's budget. This shape is pre-existing and shared with the dynamic path, but this PR extends it to a new per-request surface; the tracked "memoize the Secrets Manager client" follow-up should thread `ctx` at the same time.

3. *(minor)* A target with neither `dsn` nor `dsn_from` still errors `target %q missing dsn` — it should say "missing dsn or dsn_from" so a user who mis-indented a `dsn_from` block isn't steered away from the feature.

4. *(minor)* `password_ref` supports the `{target}` placeholder (via the shared credential resolver) but `config_ref` doesn't — a surprising asymmetry for anyone reading the credentials docs. Support it in `config_ref` or note the difference in the field doc.

5. *(nits)* The host field isn't whitespace-trimmed while the port field is — `"host": " db.example "` assembles a DSN that fails only at dial time. And the `json.Number` case in `optionalNumericField` is unreachable via `json.Unmarshal` into `map[string]any` (only a decoder with `UseNumber` produces it) — harmless, but it's a branch no input can reach.

**Verdict: LGTM once finding 1 is addressed with a doc note or an explicit follow-up commitment** — the resolver code itself is clean, well-validated, and well-tested; the concern is operational sequencing, not the implementation.

🤖 This review was generated by Claude Code (claude-fable-5) with maintainer approval.

**Comment by Kiran01bm** [comment]
> **Verdict: LGTM once finding 1 is addressed with a doc note or an explicit follow-up commitment** — the resolver code itself is clean, well-validated, and well-tested; the concern is operational sequencing, not the implementation.

yep this is known and i've added a doc note to the `DSNFrom` field that captures exactly what the review flagged: per-request assembly runs before the target-router client-cache lookup, so on a cache hit the fresh DSN is discarded — meaning secretsmanager: refs add a per-request GetSecretValue cost + availability dependency with no payoff until the DSN-aware client cache lands, while `file:` refs (the ESO case) are negligible - so it steers users to `file:` refs until that follow-up - 27c44504f856c21b183eecee0c749a54c8d4731f

**Comment by Kiran01bm** [comment]
> * _(minor)_ A target with neither `dsn` nor `dsn_from` still errors `target %q missing dsn` — it should say "missing dsn or dsn_from" so a user who mis-indented a `dsn_from` block isn't steered away from the feature.
> * _(minor)_ `password_ref` supports the `{target}` placeholder (via the shared credential resolver) but `config_ref` doesn't — a surprising asymmetry for anyone reading the credentials docs. Support it in `config_ref` or note the difference in the field doc.
> * _(nits)_ The host field isn't whitespace-trimmed while the port field is — `"host": " db.example "` assembles a DSN that fails only at dial time. And the `json.Number` case in `optionalNumericField` is unreachable via `json.Unmarshal` into `map[string]any` (only a decoder with `UseNumber` produces it) — harmless, but it's a branch no input can reach.

all the nits are now addressed as well.. - 737c9db1c63d28fa1dfc05d3730cc420d5f4aad3
